pset5/speller/dictionary.c

```diff
@@ -10,88 +10,114 @@
-// Hash function found on the internet 
-// https://en.wikipedia.org/wiki/Adler-32
-uint32_t adler32(const char *data, size_t len) 
-/* where data is the location of the data in physical memory and 
-   len is the length of the data in bytes */
-{
-    uint32_t a = 1, b = 0;
-    size_t index;
-    
-    /* Process each byte of the data in order */
-    for (index = 0; index < len; ++index)
-    {
-        a = (a + data[index]) % MOD_ADLER;
-        b = (b + a) % MOD_ADLER;
+/* Returns hash for word*/
+int hash(const char *word) {
+    int index = 0;
+    int i;
+    for (i = 0; i < strlen(word); i++) {
+        index += tolower(word[i]);
     }
-    
-    return ((b << 16) | a) % HASH_SIZE;
+    index = index % HASH_SIZE;
+    return index;  
 };
 
-/* Returns hash for word*/
-int hash(const char *word);
+int wordCount;
+bool loaded = false;
 
 // Create the node data type for the hash table
 // each element of the hash table array is a node pointer
 typedef struct node {
-    char word[LENGTH + 1];
-    struct node *next;
+    bool is_word;
+    struct node *children[CHAR_AMOUNT];
 } node;
 
-// Declare the hash table
-node *hashtable[HASH_SIZE] = {NULL};
+// Root of trie
+node *root;
 
-// TODO Make a hashtable datatype
 
 /**
  * Returns true if word is in dictionary else false.
  */
 bool check(const char *word)
 {
-    // Traverse the linked list in the hash table to find our word.
-    // Head of the linked list in the bucket we want to look through
-    node *head = hashtable[adler32(word, sizeof(word))];
+    // for each letter in input word
+      // go to corresponding element in children
+        // if NULL, word is mispelled
+        // if not NULL, move to next letter
+      // once at end of input word
+        // check if is_word true
     
-    node *cursor = head;
+    node *cursor = root;
+    int i;
     
-    while(cursor != NULL) {
-        
-        int comp = strcasecmp(cursor -> word, word);
-        
-        // if they match
-        if (comp == 0) {
-            // word is spelled correctly (in dictionary)
-            return true;
+    for (i = 0; i < strlen(word); i++) {
+        int index = letter_index(word[i]);
+        if (cursor -> children[index] == NULL) {
+            return false; 
         }
-        // if they don't match
         else {
-            // check next word
-            cursor = cursor -> next;
+            cursor = cursor -> children[index];
         }
     }
-    // If we are out here, that means we did not find the word or it
-    // was misspelled.
-    return false;
+    if (cursor -> is_word) {
+        return true;
+    }
+    else {
+        return false;
+    }
 }
 
+/* Finds a letter's place in the alphabet */
+int letter_index(char letter) {
+    int letter_index;
+    if (isupper(letter)) {
+        letter_index = letter % 65;
+    }
+    else if (islower(letter)) {
+        letter_index = letter % 97;
+    }
+    else {
+        // we have another char
+        letter_index = 27;
```

Comment:
Well... the only non-alpha character that we will be dealing with is an apostrophe, based on how they wrote speller.c